extract_feature_print.py

```diff
@@ -1,128 +1,110 @@
+        wav, sr = sf.read(wav_path)
+        assert sr == 16000
+        feats = torch.from_numpy(wav).float()
+        if feats.dim() == 2:  # double channels
+            feats = feats.mean(-1)
+        assert feats.dim() == 1, feats.dim()
+        if normalize:
+            with torch.no_grad():
+                feats = F.layer_norm(feats, feats.shape)
+        feats = feats.view(1, -1)
+        return feats
+
+
+    # HuBERT model
+    # if hubert model doesn't exist
+    if os.access(model_path, os.F_OK) == False:
+        raise Exception(f"Extracting failed because `{model_path}` does not exist.")
+
+    log.info("load model(s) from {}".format(model_path))
+    models, saved_cfg, task = checkpoint_utils.load_model_ensemble_and_task(
+        [model_path],
+        suffix="",
     )
-    exit(0)
-models, saved_cfg, task = checkpoint_utils.load_model_ensemble_and_task(
-    [model_path],
-    suffix="",
-)
-model = models[0]
-model = model.to(device)
-printt("move model to %s" % device)
-if device not in ["mps", "cpu"]:
-    model = model.half()
-model.eval()
-
-todo = sorted(list(os.listdir(wavPath)))[i_part::n_part]
-n = max(1, len(todo) // 10)  # 最多打印十条
-if len(todo) == 0:
-    printt("no-feature-todo")
-else:
-    printt("all-feature-%s" % len(todo))
-    for idx, file in enumerate(todo):
-        try:
-            if file.endswith(".wav"):
-                wav_path = "%s/%s" % (wavPath, file)
-                out_path = "%s/%s" % (outPath, file.replace("wav", "npy"))
-
-                if os.path.exists(out_path):
-                    continue
-
-                feats = readwave(wav_path, normalize=saved_cfg.task.normalize)
-                padding_mask = torch.BoolTensor(feats.shape).fill_(False)
-                inputs = {
-                    "source": feats.half().to(device)
-                    if device not in ["mps", "cpu"]
-                    else feats.to(device),
-                    "padding_mask": padding_mask.to(device),
-                    "output_layer": 9 if version == "v1" else 12,  # layer 9
-                }
-                with torch.no_grad():
-                    logits = model.extract_features(**inputs)
-                    feats = (
-                        model.final_proj(logits[0]) if version == "v1" else logits[0]
-                    )
-
-                feats = feats.squeeze(0).float().cpu().numpy()
-                if np.isnan(feats).sum() == 0:
-                    np.save(out_path, feats, allow_pickle=False)
-                else:
-                    printt("%s-contains nan" % file)
-                if idx % n == 0:
-                    printt("now-%s,all-%s,%s,%s" % (len(todo), idx, file, feats.shape))
-        except:
-            raise Exception(traceback.format_exc())
-    printt("all-feature-done")
+    model = models[0]
+    model = model.to(device)
+    log.info("move model to %s" % device)
+    if device not in ["mps", "cpu"]:
+        model = model.half()

```

Comment:
this model should be load once in an __init__ of a class.
it is not being trained, no reason to reload on each call

Reply:
We do this in may places (including inference), with `hubert_base.pt`, and also UVR and AST code (though I'm not sure if it's actually used there).

In any case, I didn't change any functionality in this PR (before, we did this in a new subprocess once for every training call).
If we want to optimize this (we have a very long loading time, so we definitely should), I think this should happen in a different PR.

Reply:
fair enough